style: Split the invalidation processing from the invalidator step.

emilio opened the pull request:
This is the first step in reusing the invalidation machinery for other stuff,
potentially including QuerySelector / QuerySelectorAll.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/18847)
<!-- Reviewable:end -->

emilio committed:
style: Split the invalidation collection from the invalidator step.

This is the first step in reusing the invalidation machinery for other stuff,
potentially including QuerySelector / QuerySelectorAll.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/data.rs, components/style/invalidation/element/collector.rs, components/style/invalidation/element/invalidator.rs, components/style/invalidation/element/mod.rs
 * @canaltinova: components/style/data.rs, components/style/invalidation/element/collector.rs, components/style/invalidation/element/invalidator.rs, components/style/invalidation/element/mod.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

emilio commented:
r? @heycam or @SimonSapin 

This is just moving code around, no behavior change (apart from introducing the `InvalidatorCollector` trait).

emilio committed:
style: Isolate all the restyling related logic in the invalidator in an InvalidationProcessor trait.

Ditto, no change in behavior.

emilio committed:
style: Move the bounds up in InvalidationProcessor.

emilio committed:
style: Unify invalidated_child with invalidated_descendants.

I think invalidated_descendants was buggy, and this fixes it.

emilio committed:
Be more precise in should_process_descendants.

emilio committed:
style: Add a way to skip the "invalidation on eager pseudo invalidates self" bit.

emilio committed:
style: Make InvalidationProcessor methods be &mut self.

This would allow querySelector / querySelectorAll to mutate the list of matched
nodes as it sees fit.

emilio commented:
I ended up working more on top of this, so please review each commit individually.

Thanks a lot!

heycam commented on components/style/invalidation/element/invalidator.rs:
```diff
@@ -384,23 +377,23 @@ where
         // but may be arbitrarily deep.
         //
         // Since we keep the traversal flags in terms of the flattened tree,
         // we need to propagate it as appropriate.
```
This comment seems to be describing what we do inside the `invalidated_descendants` implementation, which could be anything on some abritrary InvalidationProcessor.  This comment is already duplicated in the (one) `invalidated_descendants` implementation so I think we should just drop this comment here.

heycam commented on components/style/invalidation/element/invalidator.rs:
```diff
@@ -22,6 +22,11 @@ pub trait InvalidationProcessor<E>
 where
     E: TElement,
 {
+    /// Whether an invalidation that only leaves an eager pseudo-element like
```
(Not sure what you mean by "leaves".)

heycam reviewed (approved):
Looks good!

emilio commented:
@bors-servo r=heycam p=1

bors-servo commented:
:pushpin: Commit e447f81 has been approved by `heycam`

<!-- @bors-servo r=heycam e447f819a22e71d586dc25af8e03107a169d5cb1 -->

bors-servo commented:
:hourglass: Testing commit e447f819a22e71d586dc25af8e03107a169d5cb1 with merge a89a76e1fc138bd1a60ee974abfc47a4b0f69c09...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/9336), [arm32](http://build.servo.org/builders/arm32/builds/9564), [arm64](http://build.servo.org/builders/arm64/builds/9598), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9623), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/6442), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/6478), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9699), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/2119), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2133), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/6393), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/6446), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/2174), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/2160), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4992)
Approved by: heycam
Pushing a89a76e1fc138bd1a60ee974abfc47a4b0f69c09 to master...
